<center>
<img src="https://raw.githubusercontent.com/dvgodoy/PyTorch101_ODSC_Europe2020/master/images/linear_dogs.jpg" width="800">

# Основы глубинного обучения: домашнее задание 4
</center>

**ФИО:** Галиуллин Руслан Маратович

**Забавный факт о себе:**

## Общая информация

__Дата выдачи:__ 09.12.2023

__Дедлайн:__ 23:59MSK 18.12.2023

В случае этой домашки мягкий дедлан совпадает с жёстким.

## Оценивание и штрафы

Каждая из задач имеет определенную «стоимость» (указана в скобках около задачи). Максимально допустимая оценка за работу — 5.2 балла (3 основных балла и ещё 2.2 бонусных).

Задание выполняется самостоятельно. «Похожие» решения считаются плагиатом и все задействованные студенты (в том числе те, у кого списали) не могут получить за него больше 0 баллов. Если вы нашли решение какого-то из заданий (или его часть) в открытом источнике, необходимо указать ссылку на этот источник (скорее всего вы будете не единственным, кто это нашел, поэтому чтобы исключить подозрение в плагиате, необходима ссылка на источник).

In [1]:
!pip install wandb pytorch_lightning

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.1/2.1 MB 7.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 776.9/776.9 kB 10.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 190.6/190.6 kB 16.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 254.1/254.1 kB 15.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 806.1/806.1 kB 22.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.7/62.7 kB 7.8 MB/s eta 0:00:00


In [2]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from pytorch_lightning.loggers import WandbLogger
from tqdm.notebook import tqdm

In [3]:
import torch
device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')
device

device(type='cuda', index=0)

In [4]:
import wandb

wandb.login()

<IPython.core.display.Javascript object>

wandb: Appending key for api.wandb.ai to your netrc file: /root/.netrc


True

# Описание данных

Данные можно [скачать с гугл-диска.](https://drive.google.com/drive/folders/11oCcLplWtp_qm-WuEbfCFP_Mz5K_z3ps?usp=sharing) Если вы делаете задание в колабе, то строчки ниже сами скачают вам данные.



In [6]:
import gdown

url = "https://drive.google.com/drive/folders/1S3dlobf92c_qEp23VAV6WKUJrvoUeRHo?usp=sharing"
gdown.download_folder(url, quiet=True, use_cookies=False)

['/content/news_data/ria_news.tsv',
 '/content/news_data/vk_comments.tsv',
 '/content/news_data/vk_news.tsv']

Скачиваем полученные датасеты с предсказанными тегами и сантиментами

In [7]:
url = "https://drive.google.com/drive/folders/15cPQXElrR9e7Y04Bmlk8t5bopoWDrywQ?usp=drive_link"
gdown.download_folder(url, quiet=True, use_cookies=False)

['/content/data/df_comments_labeled.csv',
 '/content/data/df_comments.csv',
 '/content/data/df_vk.csv']

В таблице `ria_news.tsv`  лежат данные о новостях, вышедших на сайте РИА-НОВОСТИ с 15 марта 2018 года по 31 декабря 2018 года.

- `href` - уникальный идентификатор новости (ссылка на неё)
- `date` - дата публикации новости
- `time` - время публикации новости
- `title` - заголовок новости
- `snippet` - краткое описание новости
- `text` - текст новости
- `category` - категория новости
- `keywords` - ключевые слова (подкатегории новости)
- `shows` - счётчик с числом просмотров новости на сайте (на момент парсинга)

In [8]:
df_ria = pd.read_csv('news_data/ria_news.tsv', sep='\t')
df_ria = df_ria[~df_ria.tags.isnull()]
print(df_ria.shape)
df_ria.head()

(201708, 9)


href        date                    time  \
0  /20181231/1548961410.html  2018-12-31  31 декабря 2018, 23:52   
1  /20181231/1548961364.html  2018-12-31  31 декабря 2018, 23:19   
2  /20181231/1548961337.html  2018-12-31  31 декабря 2018, 23:12   
3  /20181231/1548961304.html  2018-12-31  31 декабря 2018, 23:09   
4  /20181231/1548961265.html  2018-12-31  31 декабря 2018, 23:07   

                                               title  \
0  Нетаньяху не собирается в отставку в случае пр...   
1  Макрон в новогоднем обращении затронул тему ре...   
2  Аарон Рэмзи проведет переговоры с пятью топ-кл...   
3  Гол Азмуна принес сборной Ирана победу над кат...   
4  Пятая ракетка мира дель Потро пропустит Открыт...   

                                             snippet  \
0  Премьер-министр Израиля Биньямин Нетаньяху не ...   
1  Результат реформ не может быть мгновенным, зая...   
2  Полузащитник лондонского "Арсенала" Аарон Рэмз...   
3  Футболисты сборной Ирана одержали победу над к...   
4  Аргентинский теннисист Хуан Мартин дель Потро ...   

                                                text category  \
0  МОСКВА, 31 дек - РИА Новости. Премьер-министр ...   В мире   
1  ПАРИЖ, 31 дек – РИА Новости. Результат реформ ...   В мире   
2  МОСКВА, 31 дек - РИА Новости. Полузащитник лон...      NaN   
3  МОСКВА, 31 дек - РИА Новости. Футболисты сборн...      NaN   
4  МОСКВА, 31 дек - РИА Новости. Аргентинский тен...      NaN   

                                     tags   shows  
0     Биньямин Нетаньяху, Израиль, В мире   728.0  
1       Эммануэль Макрон, Франция, В мире  3086.0  
2                     ФК Арсенал (Лондон)   183.0  
3  Сердар Азмун, Сборная Ирана по футболу    78.0  
4                                  Теннис    79.0

Многие новостные агенства поддерживают странички в социальных сетях. Они постят туда самые сочные сюжеты. В таблице `vk_news.tsv` лежат данные о новостях, которые РИА запостили ВКонтакте в период времени с  `2017-09-29 01:28:55` по `2019-02-01 23:13:17`.

- `id` - уникальный идентификатор поста
- `href` - ссылка на сайт (если она была указана в посте)
- `datetime` - дата и время публикации новости
- `title` - заголовок новости
- `text` - текст новости в социальной сети
- `likes` - число лайков под постом
- `comments` - число комментариев под постом

In [9]:
df_vk = pd.read_csv('news_data/vk_news.tsv', sep='\t')
df_vk['snippet'] = df_vk['text']
df_vk.drop('text', axis=1, inplace=True)
print(df_vk.shape)
df_vk.head()

(19928, 7)


id                       href             datetime  \
0  24006362  /20190201/1550280358.html  2019-02-01 23:13:17   
1  24006240  /20190201/1550268781.html  2019-02-01 22:38:41   
2  24006100  /20190201/1550282212.html  2019-02-01 21:58:52   
3  24005972  /20190202/1550283179.html  2019-02-01 21:27:06   
4  24005764  /20190201/1550262848.html  2019-02-01 20:55:54   

                                               title  likes  comments snippet  
0  В ДНР заявили о задержании диверсантов, причас...     15        28     NaN  
1  Житель Урала "заминировал" ТЦ из-за снятия  со...     32        42     NaN  
2      В Черном море нашли "потерянный флот Гитлера"     84        23     NaN  
3  В США освободили задержанную российскую актрис...     58        35     NaN  
4  Толкнувший Скабееву депутат Рады заявил о гроз...     45       145     NaN

В таблице `vk_comments.tsv` лежат комментарии к новостям.

- `id` - уникальный идентификатор комментария
- `post_id` - идентификатор новости, под которой был оставлен комментарий
- `datetime` - дата и время, когда был оставлен комментарий
- `text` - текст комментария
- `likes` - число лайков под комментарием

In [10]:
df_comments = pd.read_csv('news_data/vk_comments.tsv', sep='\t')
df_comments = df_comments[~df_comments.text.isnull()]
print(df_comments.shape)
df_comments.head()

<ipython-input-10-9f937a17e540>:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  df_comments = pd.read_csv('news_data/vk_comments.tsv', sep='\t')


(2612629, 5)


id     post_id             datetime  \
0  24006366.0  24006362.0  2019-02-01 23:14:14   
1  24006370.0  24006362.0  2019-02-01 23:15:23   
2  24006371.0  24006362.0  2019-02-01 23:16:21   
3  24006374.0  24006362.0  2019-02-01 23:16:38   
4  24006375.0  24006362.0  2019-02-01 23:16:40   

                                                text  likes  
0                                       ЧВК Вагнера?    5.0  
1                [id4710641|Евгений], выздоравливай.    3.0  
2  [id442655034|Андрей], искренне желаю этого все...    4.0  
3                         Опять про Украину новости?    1.0  
4                                      Че такое ДНР?    2.0

# А что надо сделать то?

В тетрадке вам предстоит сделать следующие шаги:

1. Вы обучите нейросеть предсказывать категорию новости
2. Вы построите предсказания для тех новостей, где мы ничего не знаем о категории
3. Вы используете уже обученный для сентимент-анализа классификатор из библиотеки `hugging face` чтобы предсказать эмоциональную окраску каждого комментария
4. Вы проведёте аналитику по новостям, а именно построите топы из самых позитивных и негативных категорий и новостей

Для первого шага вам будет дан бэйзлайн. Если вы его прогоните, у вас получится базовая модель, которая даст некоторое качество решения задачи. Вам надо будет выяснить, насколько это качество оказалось хорошим, а затем внести в код некоторые улучшения.



## Часть 1: категоризация новостей (1.2 + 2 бонусных балла)

Каждой новости в соотвествие поставлены ключевые слова. Будем считать, что эти ключевые слова — тематики новости. Нужно научиться предсказывать тематики по тексту новости. Готовые тематики у нас есть только по новостям с сайта. Они за 2018 год. По новостям из ВКонтакте у нас тематик нет. Мы собираемся их предсказать.

Новости, опубликованные ВКонтакте, отличаются от новостей с сайта тем, что у них есть только титул и короткое описание. Странно будет обучать нейросеть на длинных текстах, а потом использовать её на коротких описаниях. Мы не будем так делать. Мы попробуем обучить базовый вариант нейронной сети только на заголовках новостей. Все, кто захочет получить бонусные баллы, смогут попробовать добавить в нейросеть сниппеты (так назыают короткие описания новостей).

## 1.1 Подготовка таргета

Поработаем с таргетом. Мы будем предсказывать переменную `tags`. Давайте выясним скоько уникальных тегов существует.

In [11]:
from collections import Counter

# удалим все лишние пробелы и сделаем lowercase
df_ria['tags'] = (
    df_ria.tags.
    apply(lambda w: ','.join([item.strip() for item in  w.lower().split(',')]))
)

tags = ','.join(list(df_ria.tags.values))
tags_cnt = Counter(tags.split(','))

print(len(tags_cnt))
tags_cnt.most_common()[-20:]

13344


[('блог анны завершинской об автоспорте - блоги', 1),
 ('министерство транспорта рб', 1),
 ('министерство здравоохранения грузии', 1),
 ('палех', 1),
 ('юрий посохов (хореограф)', 1),
 ('мария александрова', 1),
 ('том бенсон', 1),
 ('абдул каюм кочай', 1),
 ('нуман куртулмуш', 1),
 ('mipim', 1),
 ('владимир попов', 1),
 ('брюно женезио', 1),
 ('роберт фицо', 1),
 ('сергей пашинский', 1),
 ('валерия гонтарева', 1),
 ('нововоронеж', 1),
 ('императорское православное палестинское общество', 1),
 ('event_poslanie_prezidenta_rf_federalnomu_sobraniju', 1),
 ('фхтр', 1),
 ('игорь честин', 1)]

Всего в выборке есть порядка 13 000 тэгов. Многие встречаются всего по разу. Давайте оставим в выборке только те тэги, которые встречаются более 30 раз.

In [12]:
target_tags = {tag for tag,cnt in tags_cnt.most_common() if cnt > 30}
len(target_tags)

1583

Закодируем теги для OHE.

In [13]:
tag2idx = dict(zip(target_tags, range(len(target_tags))))
idx2tag = {jtem: item for item,jtem in tag2idx.items()}

Почистим таргет от лишних тэгов.

In [14]:
df_ria['target_tags'] = (
    df_ria.tags.
    apply(lambda w: [tag2idx.get(item) for item in w.split(',') if item in target_tags])
)

df_ria = df_ria[df_ria.target_tags.apply(len) > 0]
df_ria.shape

(201437, 10)

In [15]:
df_ria.target_tags.values[:3]

array([list([1493, 455, 879]), list([487, 790, 879]), list([896])],
      dtype=object)

## 1.2 Подготовка текстов

Теперь займёмся предобработкой текстов. Приведём все слова к маленькому регистру и выбросим мусорные символы. В качестве токенов будем рассматривать отдельные слова.

Напомню, что мы пока что решили работать только с названиями статей. Поэтому вся предобработка применяется исключительно к ним. **Спойлер:** предобработку для сниппетов вы сделаете сами в первом же задании.

In [16]:
import nltk
nltk.download('punkt')
nltk.download('stopwords')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [17]:
import re
from nltk.tokenize import word_tokenize

def normalise_text(text):
    text = text.lower()

    # сурово регулярками выкидываем мусорные символы
    text = re.sub('[^а-яa-z0-9 ]', '', text)
    return text.strip()

df_ria['title_clean'] = df_ria.title.apply(normalise_text)

word_cnt = Counter(word_tokenize(' '.join(df_ria.title_clean.values)))
len(word_cnt)

112178

In [18]:
word_cnt.most_common()[:10]

[('в', 127323),
 ('на', 44386),
 ('с', 26150),
 ('и', 21771),
 ('о', 19948),
 ('по', 17014),
 ('россии', 13494),
 ('не', 13483),
 ('сша', 9942),
 ('за', 9881)]

Давайте почистим словарь от стоп-слов и подготовим его к использованию внутри датасета. Мы будем с помощью словаря заменять слова на индексы. Добавим в словарь несколько специальных токенов для неизвестных слов и паддингов.

In [19]:
from nltk.corpus import stopwords

stops_ru = set(stopwords.words('russian'))
len(stops_ru)

151

In [20]:
vocabulary = {
    "#PAD#": 0, "#UNK#": 1
}

k = 2
for word, _ in word_cnt.most_common():
    if word not in stops_ru:
        vocabulary[word] = k
        k += 1

In [21]:
len(vocabulary)

112030

Завернём код для создания словаря в функцию.

In [22]:
def create_vocab(text, stops_ru=stops_ru):

    word_cnt = Counter(word_tokenize(text))
    vocabulary = {
        "#PAD#": 0, "#UNK#": 1
    }

    k = 2
    for word, _ in word_cnt.most_common():
        if word not in stops_ru:
            vocabulary[word] = k
            k += 1
    return vocabulary

__[0.2 балла] Задание 1:__

- Cделайте аналогичную предобработку титулов из таблички `df_vk`. Запишите получившийся результат в столбец `title_clean` по аналогии с таблицей `df_ria`.
- Сделайте для обеих таблиц предобработку колонок со сниппетами `snippet` и запишите получившийся результат в столбец `snippet_clean`. Все пропуски заполните токеном `"#UNKN"`.

In [23]:
def normalise_text(text):
    text = text.lower()

    # сурово регулярками выкидываем мусорные символы
    text = re.sub('[^а-яa-z0-9 ]', '', text)
    return text.strip()

df_vk['title_clean'] = df_vk.title.apply(normalise_text)
df_vk.head()

id                       href             datetime  \
0  24006362  /20190201/1550280358.html  2019-02-01 23:13:17   
1  24006240  /20190201/1550268781.html  2019-02-01 22:38:41   
2  24006100  /20190201/1550282212.html  2019-02-01 21:58:52   
3  24005972  /20190202/1550283179.html  2019-02-01 21:27:06   
4  24005764  /20190201/1550262848.html  2019-02-01 20:55:54   

                                               title  likes  comments snippet  \
0  В ДНР заявили о задержании диверсантов, причас...     15        28     NaN   
1  Житель Урала "заминировал" ТЦ из-за снятия  со...     32        42     NaN   
2      В Черном море нашли "потерянный флот Гитлера"     84        23     NaN   
3  В США освободили задержанную российскую актрис...     58        35     NaN   
4  Толкнувший Скабееву депутат Рады заявил о гроз...     45       145     NaN   

                                         title_clean  
0  в днр заявили о задержании диверсантов причаст...  
1  житель урала заминировал тц изза снятия  сотов...  
2        в черном море нашли потерянный флот гитлера  
3  в сша освободили задержанную российскую актрис...  
4  толкнувший скабееву депутат рады заявил о гроз...

In [24]:
df_ria['snippet_clean'] = df_ria[~df_ria.snippet.isna()].snippet.apply(normalise_text)
df_ria.fillna({'snippet_clean': "#UNK#"}, inplace=True)

df_vk['snippet_clean'] = df_vk[~df_vk.snippet.isna()].snippet.apply(normalise_text)
df_vk.fillna({'snippet_clean': "#UNK#"}, inplace=True)

In [ ]:
df_vk.shape

(19928, 9)

## 1.3 Поставка данных

Пересечение сайта и ВК по опубликованным новостям довольно маленькое. Мы обучаем модель на данных с сайта. Предсказания мы будем строить на данных из ВК. У этих данных разная природа. В ВК описание статей и заголовки короче. Модель может хорошо показать себя на данных с новостного сайта, но сильно просесть в качестве на данных из ВК.

Давайте сохраним пересечение в отдельную табличку, чтобы на нём можно было понять, насколько сильно деградирует модель.

In [ ]:
ria_hrefs = set(df_ria.href.values)
vk_hrefs = set(df_vk.href.values)
test_hrefs = (vk_hrefs & ria_hrefs)

print('Размер отложенной выборки:', len(test_hrefs))

df = df_ria[~df_ria.href.isin(test_hrefs)]

Размер отложенной выборки: 1128


По странному совпадению (я правда не знаю почему) пересечение лежит в декабре. Мы будем его использовать как тестовую выборку.

In [ ]:
df_ria[df_ria.href.isin(test_hrefs)].date.min(), df_ria[df_ria.href.isin(test_hrefs)].date.max()

('2018-12-06', '2018-12-31')

Предположим, что мы делим выборку на обучающую и тестовую случайно. За один и тот же промежуток времени может выйти довольно большое число новостей с одинаковым заголовком. Давайте представим себе, что в тесте и трэйне есть много статей про одно и то же событие. Модель научилась на обучающей выборке хорошо его тегировать. Остальные события модель тегирует намного хуже. Метрики на тестовой выборке высокие. В следующем месяце СМИ перестают освещать это событие, в потоке новостей совершенно другие новости. Качество модели резко проседает.

Чтобы не напороться на завышенные метрики, обычно выборку дробят на обучающую и тестовую по времени. Тогда статьи из теста будут имитировать поток новых новостей, освещающих новые события.

In [ ]:
df.date.min(), df.date.max()

('2018-03-15', '2018-12-31')

__[0.2 балла] Задание 2:__ Разбейте выборку на обучающую, валидационную и тестовую. В тест возьмите весь декабрь. В валидацию октябрь и ноябрь.

In [ ]:
# your code here
# ♡ (´｡• ω •｡`)

df_test = df[(df.date >= '2018-12-01') & (df.date <= '2018-12-31')]
df_val = df[(df.date >= '2018-10-01') & (df.date < '2018-12-01')]
df_train = df[(df.date < '2018-10-01') | (df.date > '2018-12-31')]

Сформируем отложенную выборку (пересечение ВКонтакте и РИА).

In [ ]:
df_oob = df_vk[df_vk.href.isin(test_hrefs)][['href', 'title_clean']]

df_ria_oob = df_ria[df_ria.href.isin(test_hrefs)][['href', 'target_tags']]
df_oob = df_oob.set_index('href').join(df_ria_oob.set_index('href')).reset_index()
df_oob.head()

href  \
0  /20181206/1547493936.html   
1  /20181206/1547516457.html   
2  /20181206/1547520788.html   
3  /20181206/1547521406.html   
4  /20181206/1547522342.html   

                                         title_clean target_tags  
0  эксперты определили самые бюджетные экзотическ...  [587, 615]  
1  рада приняла закон расширяющий контролируемую ...  [206, 507]  
2  россия оказалась родиной древнейших титанозавр...  [1003, 58]  
3  школа в красноярске превратилась в хогвартс из...       [966]  
4  рада решила не продлевать договор о дружбе и с...  [837, 507]

Напишем датасет для поставки данных в нейросеть.


In [ ]:
from torch.utils.data import Dataset
from torch.utils.data import DataLoader

class NewsDataset(Dataset):

    def __init__(self, target, title, vocab, vocab_size, max_title_len, max_classes, snippet=None, max_snippet_len=None):

       self.vocab = {word: idx  for word,idx in vocab.items() if idx < vocab_size}
       self.max_classes = max_classes
       self.y=self.target_ohe(target)
       self.X_title = self.create_text(title, max_title_len)
       self.X_snippet = None if snippet is None else self.create_text(snippet, max_snippet_len)

    def target_ohe(self, target):
        y = torch.zeros((len(target), self.max_classes))
        for i, t in enumerate(target):
            y[[i]*len(t), t] = 1.0
        return y

    def create_text(self, texts, max_len):
        result = [ ]
        for sent in texts:
            # {#PAD: 0, #UNKN: 1}
            sent_tokenize = [self.vocab.get(item, 1) for item in word_tokenize(sent)]

            # приводим все тексты к max_len
            if len(sent_tokenize) >= max_len:
                sent_tokenize = sent_tokenize[:max_len]
            else:
                sent_tokenize += [0] * (max_len - len(sent_tokenize))
            result.append(sent_tokenize)
        return torch.tensor(result, dtype=torch.int)

    def __len__(self):
        return len(self.X_title)

    def __getitem__(self, idx):
        if self.X_snippet is not None:
          return (self.X_title[idx, :], self.X_snippet[idx, :], self.y[idx])
        return (self.X_title[idx, :], self.y[idx])


__[0.2 балла] Задание 3:__ Сейчас датасет умеет работать только с полем `title_clean`. Давайте сделаем этот датасет более многофукнциональным и добавим в него возможность добавить в обработку данных сниппет.

1. Внутри датасета `snippet` надо обработать точно также как и `title`.
2. Если `snippet=None`, датасет должен вернуть два объекта: `X_title, y`. В обратном случае датасет должен вернуть три объекта.

**Важно:** Весь код ниже работает сейчас без сниппета. Он не должен развалиться от того, что сниппет в нём нигде не указан.

In [ ]:
### ╰( ͡° ͜ʖ ͡° )つ▬▬ι═══════  bzzzzzzzzzz
# will the code be with you

# Внесите в код датасета выше свои правки

Объявим датасеты, оставим в словаре 30 000 самых частотных слов. Будем смотреть на титулы максимальной длины 20.

In [ ]:
CLASSES_NUM = len(idx2tag)
VOCAB_SIZE = 10000
MAX_TITLE_LEN = 20
MAX_SNIPPET_LEN = 40
USE_SNIPPET = True


# словарь создаем по всей выборке
vocabulary = create_vocab(' '.join(df_ria.title_clean.values)) #np.concatenate((df_ria.title_clean.values, df_ria.snippet.values)
# объявляем датасеты
if USE_SNIPPET:
  train_dataset = NewsDataset(df_train.target_tags.values, df_train.title_clean.values, vocabulary, VOCAB_SIZE, MAX_TITLE_LEN, CLASSES_NUM, df_train.snippet_clean.values, MAX_SNIPPET_LEN)
  val_dataset = NewsDataset(df_val.target_tags.values, df_val.title_clean.values, vocabulary, VOCAB_SIZE, MAX_TITLE_LEN, CLASSES_NUM, df_val.snippet_clean.values, MAX_SNIPPET_LEN)
  test_dataset = NewsDataset(df_test.target_tags.values, df_test.title_clean.values, vocabulary, VOCAB_SIZE, MAX_TITLE_LEN, CLASSES_NUM, df_test.snippet_clean.values, MAX_SNIPPET_LEN)
else:
  train_dataset = NewsDataset(df_train.target_tags.values, df_train.title_clean.values, vocabulary, VOCAB_SIZE, MAX_TITLE_LEN, CLASSES_NUM)
  val_dataset = NewsDataset(df_val.target_tags.values, df_val.title_clean.values, vocabulary, VOCAB_SIZE, MAX_TITLE_LEN, CLASSES_NUM)
  test_dataset = NewsDataset(df_test.target_tags.values, df_test.title_clean.values, vocabulary, VOCAB_SIZE, MAX_TITLE_LEN, CLASSES_NUM)


In [ ]:
train_dataloader = DataLoader(train_dataset, shuffle=True, batch_size=64, num_workers=2)
val_dataloader = DataLoader(val_dataset, shuffle=False, batch_size=4096, num_workers=2)

## 1.4 Архитектуры

Соберём базовую архитектуру для обучения.

In [ ]:
from torch import nn
import torch.nn.functional as F

class SimpleClassifier(nn.Module):

    def __init__(self, vocab_size, embedding_dim, output_dim, num_texts=1):
        super().__init__()
        self.embedding = nn.Embedding(vocab_size, embedding_dim)
        self.fc = nn.Linear(embedding_dim * num_texts, output_dim)

    def forward(self, title, snippet=None):
        embedded = self.embedding(title)
        embedded = embedded.mean(dim=1)

        if snippet is not None:
          embedded_snippet = self.embedding(snippet)
          embedded_snippet = embedded_snippet.mean(dim=1)
          embedded = torch.cat((embedded, embedded_snippet), dim=1)
        else:
          embedded = torch.cat((embedded, embedded.new_zeros(embedded.shape)), dim=1)
        return self.fc(embedded)

Соберём в `pytorch_lightning` модуль для обучения нейронки.

In [ ]:
import pytorch_lightning as pl

class TrainLightningModule(pl.LightningModule):
    def __init__(self, model, learning_rate, criterion):
        super().__init__()
        self.model = model
        self.criterion = criterion
        self.learning_rate = learning_rate

    def forward(self, title, snippet=None):
        result = self.model(title, snippet)
        return result

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=self.learning_rate)
        return optimizer

    def training_step(self, train_batch, batch_idx):
        data = train_batch
        logits = self.model(*data[:-1])
        loss = self.criterion(logits, data[-1])
        self.log(
            "train_loss", loss, prog_bar=True
        )
        return loss

    def validation_step(self, val_batch, batch_idx):
        data = val_batch
        logits = self.model(*data[:-1])
        loss = self.criterion(logits, data[-1])
        self.log(
            "val_loss", loss, prog_bar=True
        )
        return loss

Обучим модель.

In [ ]:
EMBEDDING_DIM = 300
EPOCHS = 5
LR = 1e-3

wandb_logger = WandbLogger(log_model='all')
project = "hw_04_task_01"
if USE_SNIPPET:
  project += "_snippet"
wandb.init(project=project)

model_baseline = SimpleClassifier(VOCAB_SIZE, EMBEDDING_DIM, CLASSES_NUM, 2 if USE_SNIPPET else 1).to(device)
criterion = torch.nn.CrossEntropyLoss()

train_module = TrainLightningModule(model_baseline, LR, criterion)

trainer = pl.Trainer(accelerator="gpu", logger=wandb_logger, max_epochs=EPOCHS)
trainer.fit(train_module, train_dataloader, val_dataloader)
model_baseline = model_baseline.to(device)

INFO:pytorch_lightning.utilities.rank_zero:GPU available: True (cuda), used: True
INFO:pytorch_lightning.utilities.rank_zero:TPU available: False, using: 0 TPU cores
INFO:pytorch_lightning.utilities.rank_zero:IPU available: False, using: 0 IPUs
INFO:pytorch_lightning.utilities.rank_zero:HPU available: False, using: 0 HPUs
/usr/local/lib/python3.10/dist-packages/pytorch_lightning/loggers/wandb.py:389: There is a wandb run already in progress and newly created instances of `WandbLogger` will reuse this run. If this is not desired, call `wandb.finish()` before instantiating `WandbLogger`.
INFO:pytorch_lightning.accelerators.cuda:LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
INFO:pytorch_lightning.callbacks.model_summary:
  | Name      | Type             | Params
-----------------------------------------------
0 | model     | SimpleClassifier | 4.0 M 
1 | criterion | CrossEntropyLoss | 0     
-----------------------------------------------
4.0 M     Trainable params
0         Non-trainable par

Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

INFO:pytorch_lightning.utilities.rank_zero:`Trainer.fit` stopped: `max_epochs=5` reached.


__[0.2 балла] Задание 4:__  Модель выше работает только с титулом.

- Залоггируйте её обучение на [WandB](https://wandb.ai/).
- Соберите архитектуру, которая будет принимать на вход не только титул, но ещё и сниппет. В этой архитектуре должно происходить следующее:

1. Общий слой `nn.Embedding` применяется к сниппету и титулу параллельно.
2. Происходит усреднее по текстам.
3. Вектора конкатятся в один длины 600
4. Линейный слой делает классификацию

Обучите эту модель. Сравните траектории обучения на WandB. Прикрепите ссылку на дашборд либо скришот к тетрадке.

Даталоадеры придётся объявить заново с учётом сниппетов. Правда ли, что она бьёт на валидационной выборке модель, обученную только на титулах статей?

[Запуск без сниппетов](https://wandb.ai/ruslan_iad/hw_04_task_01/runs/ph74zfqw)

[Запуск со сниппетами](https://wandb.ai/ruslan_iad/hw_04_task_01_snippet/runs/emlpy43l)

Правда ли, что она бьёт на валидационной выборке модель, обученную только на титулах статей?
Немного лучше 10.4 против 10.7

## 1.5 Инференс и оценка качества моделей

Для каждой новости нам надо предсказывать несколько тэгов. То есть в нашем случае настоящее значение таргета это множество из тэгов $y_i = [tag1, tag2, tag3]$. Прогноз модели также множество из тэгов $\hat y_i = [tag1, tag4]$.

Будем считать метрики качества следующим образом (под $|A|$ имеется в виду мощность множества, то есть число элементов в нём):

$$
Precision = \frac{1}{n} \sum_{i = 1}^n \frac{|y_i \cap \hat{y}_i|}{|\hat{y}_i|}
$$

$$
Recall = \frac{1}{n} \sum_{i = 1}^n \frac{|y_i \cap \hat{y}_i|}{|y_i|}
$$

Также можно считать аналог Accuracy, но это не самая удачная идея, так как у нас в выборке огромное число нулей и эта метрика при любом разумном пороге для принятия решения будет очень высокой.

$$
Exact Match = \frac{1}{n} \cdot \frac{1}{k} \sum_{i = 1}^n \sum_{k=1}^K [y_{ij} = \hat{y}_{ij}]
$$

In [ ]:
def precision(target, y_pred):
    num = ((y_pred == 1) & (target == 1)).sum(dim=1)
    denum = (y_pred == 1).sum(dim=1)
    return (num/(denum + 1e-5)).mean().item()

def recall(target, y_pred):
    num = ((y_pred == 1) & (target == 1)).sum(dim=1)
    denum = (target == 1).sum(dim=1)
    return (num/(denum + 1e-5)).mean().item()

def exact_match(target, y_pred):
    return (1.*(y_pred == target)).mean().item()

Построим прогноз на тестовой выборке.

In [ ]:
test_dataloader = DataLoader(test_dataset, shuffle=False, batch_size=test_dataset.__len__())

for data in test_dataloader:
    data[0], data[1] = data[0].to(device), data[1].to(device)
    if len(data) == 3:
      data[2] = data[2].to(device)
    logit = model_baseline(*data[:-1])
    pred_prob = F.softmax(logit, dim=1)
    target = data[-1]
assert pred_prob.shape[0] == test_dataset.__len__()

Теперь выбирая различное значение порога, мы можем получать разные предсказания. Если взять очень большое значение порога, то метрики сильно просядут, так как во многих документах никакого прогноза не будет построено вообще.

In [ ]:
TRESHOLD = 0.01
y_pred = 1*(pred_prob > TRESHOLD)

print('Exact Match:', exact_match(target, y_pred))
print('Precision:', precision(target, y_pred))
print('Recall:', recall(target, y_pred))

Exact Match: 0.9952263832092285
Precision: 0.27474164962768555
Recall: 0.7588459849357605


In [ ]:
TRESHOLD = 0.05
y_pred = 1*(pred_prob > TRESHOLD)
print('Exact Match:', exact_match(target, y_pred))
print('Precision:', precision(target, y_pred))
print('Recall:', recall(target, y_pred))

Exact Match: 0.9983134269714355
Precision: 0.5165744423866272
Recall: 0.5933136343955994


In [ ]:
TRESHOLD = 0.9
y_pred = 1*(pred_prob > TRESHOLD)

print('Exact Match:', exact_match(target, y_pred))
print('Precision:', precision(target, y_pred))
print('Recall:', recall(target, y_pred))

Exact Match: 0.998319685459137
Precision: 0.008524894714355469
Recall: 0.007983829826116562


Дальше мы будем строить довольно много прогнозов. Давайте напишем код для их строительства в виде функции. Обратите внимание, что на модели со снипетом она упадёт. Когда вы доберётесь до строительства прогнозов, функцию придётся немного модернизировать.

In [ ]:
def get_predict(model, dataset):
    dataloader = DataLoader(dataset, shuffle=False, batch_size=dataset.__len__())

    for data in dataloader:
        data[0], data[1] = data[0].to(device), data[1].to(device)
        if len(data) == 3:
            data[2] = data[2].to(device)
        logit = model(*data[:-1])
        pred_prob = F.softmax(logit, dim=1)

    assert pred_prob.shape[0] == dataset.__len__()
    return pred_prob, data[-1]

__[0.2 балла] Задание 5:__ Какая метрика для нас в этой задаче важнее? Точность или полнота? Почему?

__ваш ответ:__ полнота, если человек захочит узнать новости по спорту, то важно, чтобы он получил все известные новости из мира спорта, даже если там иногда будет попадать политика. Потому что люди быстро пролестнут новость не про спорт, но обидно будет не увидеть новость о смерти пеле, когда все его друзья увидели. Но при этом я не хочу брать тупо везде все новости, поэтому точность должа быть на нормальном уровне, типо в чем смысл тегов, если там тупо все новости. Возьмем, чтобы пресижн был хотя бы 0.6

- Напишите функцию, которая будет подбирать оптимальное значение порога, оптимизирующее выбранную вами метрику.
- Подберите значение порога на валидационной выборке.
- Сравните модель со сниппетами и без сниппетов, используя выбранную вами метрику при оптимальном значении порога на тестовой выборке.
- Какая из них оказалась лучше?

In [ ]:
def find_best_treshold(model, dataset, step=0.001, precision_bound=0.6):
  pred_prob, target = get_predict(model, dataset)
  max_recall = -1
  chosen_precision = 0
  best_treshold = 0
  for treshold in range(1, 800):
    treshold *= step
    y_pred = 1*(pred_prob > treshold)
    precision_val = precision(target, y_pred)
    if precision_val > precision_bound:
      chosen_precision = precision_val
      max_recall = recall(target, y_pred)
      best_treshold = treshold
      break
  return max_recall, chosen_precision, best_treshold

max_recall, chosen_precision, best_treshold = find_best_treshold(model_baseline, val_dataset)

In [ ]:
max_recall, chosen_precision, best_treshold

(0.6875284314155579, 0.6032659411430359, 0.039)

__[0.2 балла] Задание 6:__  Постройте прогнозы для отложенной выборки, которая представляет из себя пересечение сайта РИА-новостей и ВКонтакте. Проседает ли на ней качество модели? Насколько сильно?

In [ ]:
df_oob.head()

href  \
0  /20181206/1547493936.html   
1  /20181206/1547516457.html   
2  /20181206/1547520788.html   
3  /20181206/1547521406.html   
4  /20181206/1547522342.html   

                                         title_clean target_tags  
0  эксперты определили самые бюджетные экзотическ...  [587, 615]  
1  рада приняла закон расширяющий контролируемую ...  [206, 507]  
2  россия оказалась родиной древнейших титанозавр...  [1003, 58]  
3  школа в красноярске превратилась в хогвартс из...       [966]  
4  рада решила не продлевать договор о дружбе и с...  [837, 507]

In [ ]:
oob_dataset = NewsDataset(df_oob.target_tags.values, df_oob.title_clean.values, vocabulary, VOCAB_SIZE, MAX_TITLE_LEN, CLASSES_NUM)
pred_prob, target = get_predict(model_baseline, oob_dataset)

TRESHOLD = best_treshold
y_pred = 1*(pred_prob > TRESHOLD)
print('Exact Match:', exact_match(target, y_pred))
print('Precision:', precision(target, y_pred))
print('Recall:', recall(target, y_pred))

Exact Match: 0.9984349012374878
Precision: 0.387563019990921
Recall: 0.2616809010505676


Без сниппетов: полнота просела на 0.5, точность заметно сильнее: 0.6 против 0.456

Со сниппетами обученную модель нельзя использовать, потому что в oob_dataset нет инфы о них, а для модели это важные фичи

## 1.6 Бонусное задание

Давайте модернизируем наши архитектуры настолько, насколько это возможно.

__[0.5 балла]__ Попробуйте собрать более большую архитектуру. Например, сразу после слоя эмбеддингов вы можете попробовать добавить свёрточные слои (`Conv1D` свёртки). Поиграйте с оптимизатором и тп.

Опишите результаты своих экспериментов ниже. Расскажите, что конкретно вы делали и удалось ли вам улучшить качество модели. Все траектории обучения залоггируйте на WandB.   

__Ваш лог экспериментов:__

-
-
-

In [ ]:
# your code here
# (⊙_⊙)

__[0.5]__ Скачайте с сайта [Rusvectores](https://rusvectores.org/ru/models/) любые новостные word2vec эмбединги. Возьмите из модели эмбеддинги для всех слов, которые встречаются вв вашем словаре и добавьте их в модель первым слоем. Заморозьте этот слой и не обновляйте в нём веса. Если у вас в словаре есть слово, но его нет среди предобученных эмбеддингов, замените его на токен `#UNKN`.

__Ваш лог экспериментов:__

-
-
-

In [ ]:
# your code here
# =^･ｪ･^=

__[1 балл]__ Зафайнтьюньте трансформер для решения задачи с помощью библиотеки `hugging face`. Выбор предобученной модели кратко обоснуйте.

__Ваш лог экспериментов:__

-
-
-

In [ ]:
# your code here
# ٩(ˊ〇ˋ*)و


Сравните все обученные модели между собой на тестовой выборке.

In [ ]:
# your code here
# (⌐■_■)

## Часть 2: предсказание категорий (0.3 балла)

Возьмите датасет `df_vk` и для всех новостей из него предскажите категории с помощью лучшей, получившейся у вас модели.

In [ ]:
df_vk['target_tags'] = [[0]]*df_vk.shape[0]

vk_entire_dataset = NewsDataset(df_vk.target_tags.values, df_vk.title_clean.values, vocabulary, VOCAB_SIZE, MAX_TITLE_LEN, CLASSES_NUM, df_vk.snippet_clean.values, 40)
pred_prob, target = get_predict(model_baseline, vk_entire_dataset)
TRESHOLD = best_treshold
y_pred = 1 * (pred_prob > TRESHOLD)
res = []
for index, tensor in enumerate(y_pred):
    nonzero_indices = torch.nonzero(tensor).squeeze(dim=1)
    if len(nonzero_indices) > 0:
        res.append(nonzero_indices.cpu().numpy())
    else:
        res.append([])
df_vk['target_tags'] = res
df_vk.head()

id                       href             datetime  \
0  24006362  /20190201/1550280358.html  2019-02-01 23:13:17   
1  24006240  /20190201/1550268781.html  2019-02-01 22:38:41   
2  24006100  /20190201/1550282212.html  2019-02-01 21:58:52   
3  24005972  /20190202/1550283179.html  2019-02-01 21:27:06   
4  24005764  /20190201/1550262848.html  2019-02-01 20:55:54   

                                               title  likes  comments snippet  \
0  В ДНР заявили о задержании диверсантов, причас...     15        28     NaN   
1  Житель Урала "заминировал" ТЦ из-за снятия  со...     32        42     NaN   
2      В Черном море нашли "потерянный флот Гитлера"     84        23     NaN   
3  В США освободили задержанную российскую актрис...     58        35     NaN   
4  Толкнувший Скабееву депутат Рады заявил о гроз...     45       145     NaN   

                                         title_clean snippet_clean  \
0  в днр заявили о задержании диверсантов причаст...         #UNK#   
1  житель урала заминировал тц изза снятия  сотов...         #UNK#   
2        в черном море нашли потерянный флот гитлера         #UNK#   
3  в сша освободили задержанную российскую актрис...         #UNK#   
4  толкнувший скабееву депутат рады заявил о гроз...         #UNK#   

                   target_tags  
0  [507, 740, 772, 1138, 1345]  
1            [740, 1308, 1570]  
2            [740, 1138, 1570]  
3             [507, 740, 1313]  
4              [507, 740, 772]

На всякий случай сохраните табличку с получившимися у вас предсказаниями. Мало ли, вы не доделаете последнее задание, а потом захотите вернуться к нему. Не прогонять же обучение нейросети и инференс по второму кругу...

In [ ]:
from google.colab import drive
drive.mount('drive')

df_vk.to_csv('df_vk.csv')
!cp df_vk.csv "drive/My Drive/iad/data"

Drive already mounted at drive; to attempt to forcibly remount, call drive.mount("drive", force_remount=True).


## Часть 3: сентимент-классификатор (0.5 балла)

В этой части тетрадки нам предстоит прогнать все комментарии из ВК через сентимент-классификатор. Мы будем делать это с помощью библиотеки HuggingFace. В ней есть удобная [функциональность pipline,](https://huggingface.co/docs/transformers/pipeline_tutorial) чтобы прогонять на своих данных уже обученные модели. 🤗🤗🤗

In [ ]:
df_comments.head()

id     post_id             datetime  \
0  24006366.0  24006362.0  2019-02-01 23:14:14   
1  24006370.0  24006362.0  2019-02-01 23:15:23   
2  24006371.0  24006362.0  2019-02-01 23:16:21   
3  24006374.0  24006362.0  2019-02-01 23:16:38   
4  24006375.0  24006362.0  2019-02-01 23:16:40   

                                                text  likes  
0                                       ЧВК Вагнера?    5.0  
1                [id4710641|Евгений], выздоравливай.    3.0  
2  [id442655034|Андрей], искренне желаю этого все...    4.0  
3                         Опять про Украину новости?    1.0  
4                                      Че такое ДНР?    2.0

Для классификации комментариев на позитивные, негативные и нейтральные будем использовать [модель, о которой можно почитать по ссылке](https://huggingface.co/seara/rubert-tiny2-russian-sentiment)

Ответьте на следующие вопросы:

1. Как зовут автора модели?
2. Как выглядит архитектура модели? Насколько модель большая?
3. На каких данных обучена модель? Адекватно ли использовать её для классификации комментариев ВК?

Установим библиотеку. 🤗🤗🤗

In [ ]:
!pip3 install transformers

Разберитесь как можно прогнать модель на корпусе комментариев и сделайте это. Да, с помощью pipeline можно запустить довольно сложные модели, обученные другими людьми в пару строчек. При объявлении модели не забудьте положить её на нужный `device` 🤗🤗🤗

In [ ]:
from transformers import pipeline
from transformers.pipelines.pt_utils import KeyDataset
from tqdm.auto import tqdm

model = pipeline(model="seara/rubert-tiny2-russian-sentiment", device=device)

In [ ]:
model("ты мне нравишься")

[{'label': 'positive', 'score': 0.8401281237602234}]

Цикл для сентимент-анализа комментариев может работать довольно долго. Я крайне рекомендую вам переодически сохранять к себе на компьютер промежуточные результаты. Итоговый результат я рекомендую записать отдельным столбиком в таблицу с комментариями, а затем сохранить полученные результаты. 🤗🤗🤗

In [25]:
from google.colab import drive
drive.mount('drive')

Mounted at drive


In [26]:
# df_comments_short = df_comments[df_comments.text.str.len() <= 2048]

# Чтобы я делал, если бы не сделали базу до меня
# model_results = pd.DataFrame(model(list(df_comments_short.text)))

# df_comments['sentiment_verdict'] = model_results.label
# df_comments['sentiment_score'] = model_results.score

# df_comments_short.to_csv('df_comments_short.csv', index=False)
# !cp df_comments_short.csv "iad/data"

In [29]:
df_comments_labeled = pd.read_csv('data/df_comments_labeled.csv')

<ipython-input-29-24edc28de089>:1: DtypeWarning: Columns (3,4) have mixed types. Specify dtype option on import or set low_memory=False.
  df_comments_labeled = pd.read_csv('data/df_comments_labeled.csv')


## Часть 4: аналитика (1 балл + 0.2 бонусных)

Мы с вами огромные молодцы. Мы обучили модель для категоризации новостей, построили с её помощью прогнозы. Мы проскорили комментарии на их сентимент-окрас. Теперь давайте проанализируем новости. Описывайте полученные результаты таким образом, чтобы не получить уголовку на 5 лет за дискредитацию чего-нибудь или оскорбление чувств кого-нибудь 💜

__[0.2 балла]__ Какая доля комментариев позитивная? Какая доля комментариев негативная? Выведите 10 самых позитивных комментариев.

Выведите 10 самых негативных комментариев, поугарайте с них. Удалите их вывод из тетрадки. Никто не должен их видеть, это должно остаться только между нами. Поззитивные не удаляйте. Они пусть останутся.

In [45]:
print(f"Proportion of the positive comments: {len(df_comments_labeled[df_comments_labeled['label'] == 'positive']) / df_comments_labeled.shape[0]}")
print(f"Proportion of the negative comments: {len(df_comments_labeled[df_comments_labeled['label'] == 'negative']) / df_comments_labeled.shape[0]}")
df_comments_labeled[df_comments_labeled['label'] == 'positive'].head(10)

Proportion of the positive comments: 0.06964512875800248
Proportion of the negative comments: 0.22130989738752094


id     post_id             datetime  \
2    24006371.0  24006362.0  2019-02-01 23:16:21   
21   24006411.0  24006362.0  2019-02-01 23:29:41   
46   24006515.0  24006362.0  2019-02-02 00:15:14   
69   24006331.0  24006240.0  2019-02-01 23:02:02   
81   24006430.0  24006240.0  2019-02-01 23:35:40   
93   24006139.0  24006100.0  2019-02-01 22:09:17   
98   24006181.0  24006100.0  2019-02-01 22:22:17   
129  24006097.0  24005972.0  2019-02-01 21:58:19   
134  24006107.0  24005972.0  2019-02-01 22:00:37   
145  24006354.0  24005972.0  2019-02-01 23:11:03   

                                                  text  likes     label  \
2    [id442655034|Андрей], искренне желаю этого все...    4.0  positive   
21   [id89551382|Юрий], ты спи..Вольдемар сам все с...    2.0  positive   
46   [id4710641|Евгений], новость открой а потом ку...    0.0  positive   
69   Стрелочника нашли, Телеграм не зря заблокирова...    0.0  positive   
81   [id352507620|Jevgeni], для начала сходи сбрей ...    0.0  positive   
93                                       Золото нашли?    0.0  positive   
98                                                 Lol    0.0  positive   
129                                                Зря    0.0  positive   
134  [id495155274|Алексей], ну если ребёнок индус, ...    0.0  positive   
145  Наших граждан будут без основательно хватать и...    0.0  positive   

        score  
2    0.676261  
21   0.659249  
46   0.951252  
69   0.969304  
81   0.812432  
93   0.993587  
98   0.819801  
129  0.758657  
134  0.639224  
145  0.562267

__[0.2 балла]__ Для каждой новости из датасета посчитайте количество негативных и позитивных комментариев под ней. Сохраните эти количества в виде новых колонок.

In [94]:
count_by_post_id_by_label = df_comments_labeled.groupby(['post_id', 'label'], group_keys=False).count()[['id']].reset_index().rename(columns={'id': 'count', 'post_id':'id'}).sort_values(by=['id'])
pivot_table_vk = count_by_post_id_by_label.pivot_table(index='id', columns='label', values='count', aggfunc='sum')

merged_df_vk = df_vk.merge(pivot_table_vk, on='id', how='left')
merged_df_vk

id                                      href  \
0      24006362                 /20190201/1550280358.html   
1      24006240                 /20190201/1550268781.html   
2      24006100                 /20190201/1550282212.html   
3      24005972                 /20190202/1550283179.html   
4      24005764                 /20190201/1550262848.html   
...         ...                                       ...   
19923  17976905        /politics/20170929/1505812885.html   
19924  17976766        /politics/20170929/1505809314.html   
19925  17976601         /science/20170928/1505787196.html   
19926  17976198           /world/20170929/1505810992.html   
19927  17976037  /defense_safety/20170929/1505808562.html   

                  datetime                                              title  \
0      2019-02-01 23:13:17  В ДНР заявили о задержании диверсантов, причас...   
1      2019-02-01 22:38:41  Житель Урала "заминировал" ТЦ из-за снятия  со...   
2      2019-02-01 21:58:52      В Черном море нашли "потерянный флот Гитлера"   
3      2019-02-01 21:27:06  В США освободили задержанную российскую актрис...   
4      2019-02-01 20:55:54  Толкнувший Скабееву депутат Рады заявил о гроз...   
...                    ...                                                ...   
19923  2017-09-29 03:51:03  Политологи назвали губернаторов, которым грози...   
19924  2017-09-29 03:20:41  Захарова рассказала, какую книгу подписывал Ла...   
19925  2017-09-29 02:55:49  Современные люди появились более 300 тысяч лет...   
19926  2017-09-29 02:00:09                     В США ребенок устроил стрельбу   
19927  2017-09-29 01:28:55  В России впервые сократили число призывников н...   

       likes  comments snippet  \
0         15        28     NaN   
1         32        42     NaN   
2         84        23     NaN   
3         58        35     NaN   
4         45       145     NaN   
...      ...       ...     ...   
19923    141       114     NaN   
19924    140        50     NaN   
19925    202        69     NaN   
19926    138        51     NaN   
19927    329       241     NaN   

                                             title_clean snippet_clean  \
0      в днр заявили о задержании диверсантов причаст...         #UNK#   
1      житель урала заминировал тц изза снятия  сотов...         #UNK#   
2            в черном море нашли потерянный флот гитлера         #UNK#   
3      в сша освободили задержанную российскую актрис...         #UNK#   
4      толкнувший скабееву депутат рады заявил о гроз...         #UNK#   
...                                                  ...           ...   
19923  политологи назвали губернаторов которым грозит...         #UNK#   
19924  захарова рассказала какую книгу подписывал лав...         #UNK#   
19925  современные люди появились более 300 тысяч лет...         #UNK#   
19926                     в сша ребенок устроил стрельбу         #UNK#   
19927  в россии впервые сократили число призывников н...         #UNK#   

       negative  neutral  positive  
0          10.0     34.0       3.0  
1          11.0     29.0       2.0  
2           4.0     16.0       2.0  
3          10.0     27.0       3.0  
4          37.0     94.0       9.0  
...         ...      ...       ...  
19923       NaN      NaN       NaN  
19924       NaN      NaN       NaN  
19925       NaN      NaN       NaN  
19926       NaN      NaN       NaN  
19927       NaN      NaN       NaN  

[19928 rows x 12 columns]

id     label  count
0  18223951.0  negative     56
1  18223951.0   neutral    166
2  18223951.0  positive      9
3  18224542.0  negative     14
4  18224542.0   neutral     69
5  18224542.0  positive      1
6  18224684.0  negative     12
7  18224684.0   neutral     30
8  18224684.0  positive      2
9  18224973.0  negative      7

__[0.2 балл]__ Правда ли, что новости с большим количеством лайков получают больше негативных комментариев? А позитивных? Правда ли, что чем больше лайков, тем под новостью больше комментариев?

Постройте визуализацию, которая могла бы это проиллюстрировать.

In [ ]:
# your code here
# ┌(ಠ_ಠ)┘

__[0.2]__  Давайте построим по сентимент-окрасу комментариев топ позитивных новостей.

- Если под новостью оставлено 100 комментариев, из которых 80 позитивные, мы можем быть уверены в том, что новость была хорошо воспринята людьми.
- Если под новостью всего один комментарий и он оказался позитивным, то у нас 100% позитивных комментариев, но это вовсе не означает, что новость была воспринята хорошо.

Получается, что сортировать по доле позитивных комментариев нельзя. Давайте поступим умнее. Доля позитивных комментариев под постом -- это случайная величина. Её довольно часто моделируют с помощью бета-распределения. Если это случайная величина, мы можем построить для неё предиктивный интервал. Пусть $u$ - число позитивных комментариев, а $d$ - нейтральных и негативных.  Пусть

$$
a = 1 + u \qquad b = 1 + d.
$$

Тогда нижняя граница $95\%$ предиктивного интервала для доли будет вычисляться по такой формуле:

$$
\frac{a}{a + b} - 1.65 \cdot \sqrt{\frac{a \cdot b}{(a + b)^2 \cdot (a + b + 1)}}
$$

Если под новостью был всего один комментарий, у такой случайной величины будет высокая дисперсия. Это означаen, что штука, которую мы вычитаем из доли, окажется высокой. Левая граница интервала окажется маленькой и мы не поднимем комментарий в нашем топе наверх. Фактически мы делаем сортировку по квантилю уровня $0.05$ вместо среднего.

От вас требуется вбить эту формулу, сделать сортировку и вывести на экран топ позитивных новостей. Подробнее про то, откуда берётся эта формула можно почитать [в этой книге.](https://disk.yandex.ru/i/Ctd08bTwC9eI3g) Ищите 4 главу, страницу 140.

In [ ]:
# your code here
# ฅ^•ﻌ•^ฅ

Построили? Срочно пришлите свою любимую позитивную  новость в общий чат!!!

Топ негативных новостей строить не будем. Вокруг итак слишком много негатива 😻😻😻

__[0.2]__ Какие категории новостей оказались самыми позитивными? Придумайте способ найти такие категории и опишите его тут.

__Ответ:__

In [ ]:
# your code here
# ٩(⁎❛ᴗ❛⁎)۶

__[0.2 бонусных]__ Проанализируйте, как температура комментария (вероятность того, что он негативный) зависит от длины трэда (число комментариев под новостью)? Значима ли эта взаимосвязь? Если вам для проверки этого хочется построить линейную регрессию, не сдерживайтесь.

In [ ]:
# your code here
# ヾ(๑╹◡╹)ﾉ